### ISSUE TITLE
<+> does not work after `liftAuthorizedService`

### ISSUE THREAD

**Issue by ioganegambaputifonguser** [jmcardon/tsec#133]
I have 
```
val Auth: SecuredRequestHandler[IO, Int, User, AugmentedJWT[HMACSHA256, Int]] =
    SecuredRequestHandler(jwtStatefulAuth)
```
and 2 roles
```
  val AdminRequired = BasicRBAC[IO, Role, User, AugmentedJWT[HMACSHA256, Int]](Administrator)
  val ReaderRequired = BasicRBAC[IO, Role, User, AugmentedJWT[HMACSHA256, Int]](Administrator, ReadAllUser)
```
try to compose two services which belongs to different permessions

```
val service1: HttpService[IO] = Auth.liftAuthorizedService(AdminRequired, TSecAuthService {
    case request @ GET -> Root / "s1" asAuthed user =>
      Ok("admin only")
  })

  val service2: HttpService[IO] = Auth.liftAuthorizedService(ReaderRequired, TSecAuthService {
    case request @ GET -> Root / "s2" asAuthed user =>
      Ok("admin and reader")
  })

val service = service1 <+> service2
```
service1 works, service2 returns 404 in any case. if i swap it 
```
val service = service2 <+> service1
```
service2 works, service1 returns 404

If I remove `Auth`, the compose works with the pure Http4s HttpService[IO]

**Comment by rsoeldner** [comment]
@ioganegambaputifonguser Which version's do you use ?

**Comment by Yaneeve** [comment]
The version is 0.0.1-M9 (@ioganegambaputifonguser is a colleague)

**Comment by jmcardon** [comment]
@Yaneeve So we have the behavior where, to protect from spidering (which has been a really confusing default, I apologize), we make authenticated services consume all (i.e it will never hit the rightmost service).


To get over this (and in general, to use authorization nicely, which we should document but I'm currently @ NE Scala with not much time), you could redo your services like this:

```scala
val service1: TsecAuthService[F, A, B] = TSecAuthService(AdminRequired) {
    case request @ GET -> Root / "s1" asAuthed user =>
      Ok("admin only")
  }

  val service2:  TsecAuthService[F, A, B] = sSecAuthService(ReaderRequired) {
    case request @ GET -> Root / "s2" asAuthed user =>
      Ok("admin and reader")
  }

val liftedServices = Auth.liftAuthorizedblahblah(service1 <+> service2)
```

I am very aware that this is not well documented and could be better, but my time is constrained right now  so I do not have time to make a sample project/update the docs at least within this week.

**Comment by jmcardon** [comment]
Did this solve your issues? @Yaneeve @ioganegambaputifonguser?

**Comment by ioganegambaputifonguser** [comment]
With this code

```
val pf: PartialFunction[SecuredRequest[IO, Int, AugmentedJWT[HMACSHA256, Int]], IO[Response[IO]]] = {
    case request @ GET -> Root / "s1" asAuthed user =>
      Ok("admin only")
  }

  val service1 = TSecAuthService.withAuthorization(AdminRequired)(pf, _: (Response[IO], AugmentedJWT[HMACSHA256, Int]) => OptionT[IO, Response[IO]])

  val service2 = TSecAuthService.withAuthorization(ReaderRequired) {
    case request @ GET -> Root / "s2" asAuthed user =>
      Ok("reader only")
  }
```
I got

```
Error:(20, 34) ambiguous reference to overloaded definition,
both method withAuthorization in object TSecAuthService of type [A, I, F[_]](auth: tsec.authorization.Authorization[F,I,A])(pf: PartialFunction[tsec.authentication.SecuredRequest[F,I,A],F[org.http4s.Response[F]]], andThen: (org.http4s.Response[F], A) => cats.data.OptionT[F,org.http4s.Response[F]])(implicit F: cats.Monad[F])tsec.authentication.TSecAuthService[I,A,F]
and  method withAuthorization in object TSecAuthService of type [A, I, F[_]](auth: tsec.authorization.Authorization[F,I,A])(pf: PartialFunction[tsec.authentication.SecuredRequest[F,I,A],F[org.http4s.Response[F]]])(implicit F: cats.Monad[F])tsec.authentication.TSecAuthService[I,A,F]
match argument types (tsec.authorization.BasicRBAC[cats.effect.IO,kombucha.auth.AuthHelpers.Role,kombucha.auth.AuthHelpers.User,tsec.authentication.AugmentedJWT[tsec.mac.imports.HMACSHA256,Int]])
  val service1 = TSecAuthService.withAuthorization(AdminRequired)(pf, _: (Response[IO], AugmentedJWT[HMACSHA256, Int]) => OptionT[IO, Response[IO]])

Error:(22, 34) ambiguous reference to overloaded definition,
both method withAuthorization in object TSecAuthService of type [A, I, F[_]](auth: tsec.authorization.Authorization[F,I,A])(pf: PartialFunction[tsec.authentication.SecuredRequest[F,I,A],F[org.http4s.Response[F]]], andThen: (org.http4s.Response[F], A) => cats.data.OptionT[F,org.http4s.Response[F]])(implicit F: cats.Monad[F])tsec.authentication.TSecAuthService[I,A,F]
and  method withAuthorization in object TSecAuthService of type [A, I, F[_]](auth: tsec.authorization.Authorization[F,I,A])(pf: PartialFunction[tsec.authentication.SecuredRequest[F,I,A],F[org.http4s.Response[F]]])(implicit F: cats.Monad[F])tsec.authentication.TSecAuthService[I,A,F]
match argument types (tsec.authorization.BasicRBAC[cats.effect.IO,kombucha.auth.AuthHelpers.Role,kombucha.auth.AuthHelpers.User,tsec.authentication.AugmentedJWT[tsec.mac.imports.HMACSHA256,Int]])
  val service2 = TSecAuthService.withAuthorization(ReaderRequired) {
```
for both of them

**Comment by jmcardon** [comment]
Try:

`service1: TsecAuthService[IO, User, AugmentedJWT[HMACSHA256, Int]]` instead of annotating the partial function. Scalac has inferencr issues around this many type params for some reason

**Comment by ioganegambaputifonguser** [comment]
Thank you so much for the support, but unfortunately I get the same error

What I have for now
```
val service1: TSecAuthService[User, AugmentedJWT[HMACSHA256, Int], IO] =
    TSecAuthService.withAuthorization[AugmentedJWT[HMACSHA256, Int], User, IO](AdminRequired) {
      case request @ GET -> Root / "s1" asAuthed user =>
        Ok("admin only")
    }
```
error
```
/Users/theodor/work/kombucha/src/main/scala/kombucha/company/CompanyService.scala:18:38: ambiguous reference to overloaded definition,
[error] both method withAuthorization in object TSecAuthService of type (auth: tsec.authorization.Authorization[cats.effect.IO,kombucha.auth.AuthHelpers.User,tsec.authentication.AugmentedJWT[tsec.mac.imports.HMACSHA256,Int]])(pf: PartialFunction[tsec.authentication.SecuredRequest[cats.effect.IO,kombucha.auth.AuthHelpers.User,tsec.authentication.AugmentedJWT[tsec.mac.imports.HMACSHA256,Int]],cats.effect.IO[org.http4s.Response[cats.effect.IO]]], andThen: (org.http4s.Response[cats.effect.IO], tsec.authentication.AugmentedJWT[tsec.mac.imports.HMACSHA256,Int]) => cats.data.OptionT[cats.effect.IO,org.http4s.Response[cats.effect.IO]])(implicit F: cats.Monad[cats.effect.IO])tsec.authentication.TSecAuthService[kombucha.auth.AuthHelpers.User,tsec.authentication.AugmentedJWT[tsec.mac.imports.HMACSHA256,Int],cats.effect.IO]
[error] and  method withAuthorization in object TSecAuthService of type (auth: tsec.authorization.Authorization[cats.effect.IO,kombucha.auth.AuthHelpers.User,tsec.authentication.AugmentedJWT[tsec.mac.imports.HMACSHA256,Int]])(pf: PartialFunction[tsec.authentication.SecuredRequest[cats.effect.IO,kombucha.auth.AuthHelpers.User,tsec.authentication.AugmentedJWT[tsec.mac.imports.HMACSHA256,Int]],cats.effect.IO[org.http4s.Response[cats.effect.IO]]])(implicit F: cats.Monad[cats.effect.IO])tsec.authentication.TSecAuthService[kombucha.auth.AuthHelpers.User,tsec.authentication.AugmentedJWT[tsec.mac.imports.HMACSHA256,Int],cats.effect.IO]
[error] match argument types (tsec.authorization.BasicRBAC[cats.effect.IO,kombucha.auth.AuthHelpers.Role,kombucha.auth.AuthHelpers.User,tsec.authentication.AugmentedJWT[tsec.mac.imports.HMACSHA256,Int]])
[error]     TSecAuthService.withAuthorization[AugmentedJWT[HMACSHA256, Int], User, IO](AdminRequired) {
```

**Comment by jmcardon** [comment]
Alright, yeah, this is an actual issue with method inference we need to fix.

Thank you for reporting. If I remove the second `withAuthorization` method that's totally unused, it compiles.

For now, I can give you a workaround to this:

```scala
    def withAuthorization[I, A, F[_]](auth: tsec.authorization.Authorization[F, I, A])(
        pf: PartialFunction[SecuredRequest[F, I, A], F[Response[F]]]
    )(implicit F: Monad[F]): TSecAuthService[I, A, F] =
      Kleisli { req: SecuredRequest[F, I, A] =>
        auth
          .isAuthorized(req)
          .flatMap(_ => pf.andThen(OptionT.liftF(_)).applyOrElse(req, Function.const(OptionT.none)))
      }
```

This will do what you want it to do. I'll make sure to remove the extra method with the `andThen` in a PR today, as this is clearly impacting inference negatively.

**Comment by ioganegambaputifonguser** [comment]
it will work but unfortunately I have to say that we came from where we started
```
def liftAuthorizedService(
      authorization: Authorization[F, User, Auth],
      service: TSecAuthService[User, Auth, F],
      onNotAuthenticated: Request[F] => F[Response[F]] = defaultNotAuthenticated
  )
```
does not accept `TSecAuthService <+> TSecAuthService`, it receives only one based on provided authorisation.
I tried `val service = Auth.liftService(service1) <+> Auth.liftService(service2)` but it also seems consumes everything in the first Auth and I get 404 from the service2 :(

**Comment by jmcardon** [comment]
@ioganegambaputifonguser this is why I am saying, please add the method in my comment above to a helper of some sort, then use that, while I fix this in a PR. i.e

```scala
object TSecAuthHelper {
    def withAuthorization[I, A, F[_]](auth: tsec.authorization.Authorization[F, I, A])(
        pf: PartialFunction[SecuredRequest[F, I, A], F[Response[F]]]
    )(implicit F: Monad[F]): TSecAuthService[I, A, F] =
      Kleisli { req: SecuredRequest[F, I, A] =>
        auth
          .isAuthorized(req)
          .flatMap(_ => pf.andThen(OptionT.liftF(_)).applyOrElse(req, Function.const(OptionT.none)))
      }
}

val service1: TSecAuthService[User, AugmentedJWT[HMACSHA256, Int], IO] =
    TSecAuthHelper.withAuthorization(AdminRequired) {
      case request @ GET -> Root / "s1" asAuthed user =>
        Ok("admin only")
    }
```

I don't mean this as a long term fix, just for now, while I create a PR to fix this issue.

**Comment by ioganegambaputifonguser** [comment]
Yes yes, I got it
It works
But the problem is here where we actually started, how to compose service1 and service2 after creation? 
Now I have 
`service1: TSecAuthService[User, AugmentedJWT[HMACSHA256, Int], IO]` and `service2: TSecAuthService[User, AugmentedJWT[HMACSHA256, Int], IO]` with this temporary solution.
As I said `val service = Auth.liftService(service1) <+> Auth.liftService(service2)` gives 404 from service2.

The `Auth.liftAuthorizedblahblah(service1 <+> service2)` does not work here also because It works with the only one `TSecAuthService` with the concrete Roles set, you can not pass `(service1 <+> service2`.

**Comment by jmcardon** [comment]
If you apply authorization on the individual services, you don't need to apply it from the handler.

Use `.liftService` and it should work just fine.

Something like:

```scala
  val auth1: BasicRBAC[IO, Role, User, Int] = BasicRBAC(Administrator: Role)
  val auth2: BasicRBAC[IO, Role, User, Int] = BasicRBAC(Customer)

  val s: TSecAuthService[User, Int, IO] =
    TSecAuthService.withAuthorization(auth1) {
      case GET -> Root asAuthed u =>
        Ok()
    }

  val q: TSecAuthService[User, Int, IO] =
    TSecAuthService.withAuthorization(auth2) {
      case GET -> Root asAuthed u =>
        Ok()
    }

  val composedService: TSecAuthService[User, Int, IO] = s <+> q

  def myhandler: SecuredRequestHandler[IO, Int, User, Int] = ???
  myhandler.liftService(composedService)
```

**Comment by ioganegambaputifonguser** [comment]
Thank you one more time, it works with this approach
```
  val composedService: TSecAuthService[User, Int, IO] = s <+> q

  def myhandler: SecuredRequestHandler[IO, Int, User, Int] = ???
  myhandler.liftService(composedService)
```

**Comment by ioganegambaputifonguser** [comment]
@jmcardon I hope it will help https://github.com/jmcardon/tsec/pull/138
I search within the codebase, it seems they both (overloaded withAuthorization and liftAuthorizedService) not used anywhere

**Comment by jmcardon** [comment]
closed by #138 

Thank you very much @ioganegambaputifonguser
